verification/certifyKey_test.go

```diff
@@ -118,21 +579,194 @@ func testCertifyKey(d TestDPEInstance, t *testing.T) {
+
+		certChain := checkCertificateChain(t, certChainBytes)
+		t.Logf("[LOG]: Certificate chain is DER encoded")
+
+		// Validate that all X.509 fields conform with the format defined in the DPE iRoT profile
+		validateCertifyKeyCert(t, leafCert, uint32(r.Flags), r.Label[:])
+
+		// Ensure full certificate chain has valid signatures
+		// This also checks certificate lifetime, signatures as part of cert chain validation
+		validateCertChain(t, certChain, leafCert)
+
+		// TODO: When DeriveChild is implemented, call it here to add more TCIs and call CertifyKey again.
+	}
+}
+
+// Build certificate chain and calls to validateSignature on each chain.
+func validateCertChain(t *testing.T, certChain []*x509.Certificate, leafCert *x509.Certificate) {
+	t.Helper()
+	var certsToProcess []*x509.Certificate
+	if leafCert != nil {
+		t.Log("[LOG]: Validating leaf certificate chain...")
+		certsToProcess = []*x509.Certificate{leafCert}
+	} else {
+		t.Log("[LOG]: Validating intermediate certificates chains...")
+		certsToProcess = certChain
+	}
+
+	// Remove unhandled critical extensions reported by x509 but defined in spec
+	t.Log("[LOG]: Checking for unhandled critical certificate extensions unknown to DPE certificates profile spec...")
+	removeTcgDiceCriticalExtensions(t, certsToProcess)
+
+	// Remove unhandled extended key usages reported by x509 but defined in spec
+	t.Log("[LOG]: Checking for extended key usages unknown to DPE certificates profile spec...")
+	removeTcgDiceExtendedKeyUsages(t, certsToProcess)
+
+	// Build verify options
+	opts := buildVerifyOptions(t, certChain)
+
+	if leafCert != nil {
+		// Certificate chain validation for leaf
+		chains, err := leafCert.Verify(opts)
+		if err != nil {
+			// Certificate chain cannot be built from leaf to root
+			t.Errorf("[ERROR]: Error in certificate chain %s: ", err.Error())
+		}
+
+		// Log certificate chains linked to leaf
+		t.Logf("[LOG]: Chains of DPE leaf certificate.")
+		for _, chain := range chains {
+			for i, cert := range chain {
+				t.Logf("%d %s", i, (*cert).Subject)
+			}
+
+		}
+
+		// This indicates that signature validation found no errors in the DPE leaf cert chain
+		t.Logf("[LOG]: DPE leaf certificate chain validation is done")
+
+	} else {
+		// Certificate chain validation for each intermediate certificate
+		for _, cert := range certChain {
+			chains, err := cert.Verify(opts)
+			if err != nil {
+				t.Errorf("[ERROR]: Error in Certificate Chain of %s: %s", cert.Subject, err.Error())
+			}
+
+			// Log certificate chains linked to each cetificate in chain
+			t.Logf("[LOG]: Chains of intermediate certificate.")
+			for _, chain := range chains {
+				for i, cert := range chain {
+					t.Logf("%d %s", i, (*cert).Subject)
+				}
+			}
+		}
+
+		// This indicates that signature validation found no errors each cert
+		// chain of intermediate certificates
+		t.Logf("[LOG]: Intermediate certificates chain validation is done")
+	}
```

Comment:
I would prefer to move this check to the GetCertificateChain test so that the test code is easier to follow. Suggested refactor:

```go
func validateCertChain(...) {
	// Remove unhandled critical extensions reported by x509 but defined in spec
	removeTcgDiceCriticalExtensions(t, certsToProcess)

	// Remove unhandled extended key usages reported by x509 but defined in spec
	removeTcgDiceExtendedKeyUsages(t, certsToProcess)

	// Build verify options
	opts := buildVerifyOptions(t, certChain)
		
         // Certificate chain validation for each intermediate certificate
	for _, cert := range certChain {
		chains, err := cert.Verify(opts)
		if err != nil {
			t.Errorf("[ERROR]: Error in Certificate Chain of %s: %s", cert.Subject, err.Error())
		}
		// Log certificate chains linked to each cetificate in chain
		t.Logf("[LOG]: Chains of intermediate certificate.")
		for _, chain := range chains {
			for i, cert := range chain {
				t.Logf("%d %s", i, (*cert).Subject)
			}
		}
	}

	// This indicates that signature validation found no errors each cert
	// chain of intermediate certificates
	t.Logf("[LOG]: Intermediate certificates chain validation is done")
}
```

Then the code here can be specifically for the leaf key

Reply:
The latest commit just moves this whole function to GetCertificateChain. To clarify, we still want to verify the leave cert chains to the root. I just think they can be split into two functions. GetCertificateChain test has the above code. Then CertifyKey test has

```go
func validateLeafCert(...) {
    	// Remove unhandled critical extensions reported by x509 but defined in spec
	removeTcgDiceCriticalExtensions(t, certsToProcess)

	// Remove unhandled extended key usages reported by x509 but defined in spec
	removeTcgDiceExtendedKeyUsages(t, certsToProcess)

	// Build verify options
	opts := buildVerifyOptions(t, certChain)

	// Certificate chain validation for leaf
	chains, err := leafCert.Verify(opts)
	if err != nil {
		// Certificate chain cannot be built from leaf to root
		t.Errorf("[ERROR]: Error in certificate chain %s: ", err.Error())
	}
	
        // Log certificate chains linked to leaf
	if len(chains) == 0 {
		t.Errorf("[ERROR]: certificate chain is empty")
	}
}
```

Reply:
Have split the cert chain validation function as suggested.